client-runtime/src/main/java/com/microsoft/rest/v2/protocol/HttpResponseDecoder.java

```diff
@@ -0,0 +1,291 @@
+                if (resultTypeToken.isSubtypeOf(List.class)) {
+                    final Type resultElementType = getTypeArgument(resultType);
+
+                    final List<Object> wireResponseList = (List<Object>) wireResponse;
+
+                    final int wireResponseListSize = wireResponseList.size();
+                    for (int i = 0; i < wireResponseListSize; ++i) {
+                        final Object wireResponseElement = wireResponseList.get(i);
+                        final Object resultElement = convertToResultType(wireResponseElement, resultElementType, wireType);
+                        if (wireResponseElement != resultElement) {
+                            wireResponseList.set(i, resultElement);
+                        }
+                    }
+
+                    result = wireResponseList;
+                }
+                else if (resultTypeToken.isSubtypeOf(Map.class)) {
+                    final Type resultValueType = getTypeArguments(resultType)[1];
+
+                    final Map<String, Object> wireResponseMap = (Map<String, Object>) wireResponse;
+
+                    final Set<String> wireResponseKeys = wireResponseMap.keySet();
+                    for (String wireResponseKey : wireResponseKeys) {
+                        final Object wireResponseValue = wireResponseMap.get(wireResponseKey);
+                        final Object resultValue = convertToResultType(wireResponseValue, resultValueType, wireType);
+                        if (wireResponseValue != resultValue) {
+                            wireResponseMap.put(wireResponseKey, resultValue);
+                        }
+                    }
+                } else if (resultTypeToken.isSubtypeOf(RestResponse.class)) {
+                    RestResponse<?, ?> restResponse = (RestResponse<?, ?>) wireResponse;
+                    Object wireResponseBody = restResponse.body();
+
+                    Object resultBody = convertToResultType(wireResponseBody, getTypeArguments(resultType)[1], wireType);
+                    if (wireResponseBody != resultBody) {
+                        result = new RestResponse<>(restResponse.statusCode(), restResponse.headers(), restResponse.rawHeaders(), resultBody);
+                    } else {
+                        result = restResponse;
+                    }
+                }
+            }
+        }
+
+        return result;
+    }
+
+    private TypeToken getEntityType() {
+        TypeToken token = TypeToken.of(methodParser.returnType());
+
+        if (token.isSubtypeOf(Single.class) || token.isSubtypeOf(Maybe.class) || token.isSubtypeOf(Observable.class)) {
+            token = TypeToken.of(getTypeArgument(token.getType()));
+        }
+
+        if (token.isSubtypeOf(RestResponse.class)) {
+            token = TypeToken.of(getTypeArguments(token.getType())[1]);
+        }
+
+        // TODO: unwrap OperationStatus a different way?
```

Comment:
Not extremely happy with this but also reluctant to introduce polymorphism (i.e. attaching an AzureResponseDecoder to the HttpRequest) until LROs are refactored. It might wind up being the best way though.